pinot-core/src/main/java/org/apache/pinot/core/operator/filter/BitmapCollection.java

```diff
@@ -0,0 +1,103 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.filter;
+
+import org.roaringbitmap.buffer.BufferFastAggregation;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+public class BitmapCollection {
+  private final int _numDocs;
+  private boolean _inverted;
+  private final ImmutableRoaringBitmap[] _bitmaps;
+
+  public BitmapCollection(int numDocs, boolean inverted, ImmutableRoaringBitmap... bitmaps) {
+    _numDocs = numDocs;
+    _inverted = inverted;
+    _bitmaps = bitmaps;
+  }
+
+  public boolean isInverted() {
+    return _inverted;
+  }
+
+  public ImmutableRoaringBitmap[] getBitmaps() {
+    return _bitmaps;
+  }
+
+  public BitmapCollection invert() {
+    _inverted = !_inverted;
+    return this;
+  }
+
+  public int andCardinality(BitmapCollection other) {
+    if (!_inverted) {
+      if (!other._inverted) {
+        return ImmutableRoaringBitmap.andCardinality(reduceInternal(), other.reduceInternal());
+      }
+      return ImmutableRoaringBitmap.andNotCardinality(reduceInternal(), other.reduceInternal());
+    } else {
+      if (!other._inverted) {
+        return ImmutableRoaringBitmap.andNotCardinality(other.reduceInternal(), reduceInternal());
+      }
+      return _numDocs - ImmutableRoaringBitmap.orCardinality(reduceInternal(), other.reduceInternal());
+    }
+  }
+
+  public int orCardinality(BitmapCollection other) {
+    if (!_inverted) {
+      if (!other._inverted) {
+        return ImmutableRoaringBitmap.orCardinality(reduceInternal(), other.reduceInternal());
+      }
+      ImmutableRoaringBitmap reduced = other.reduceInternal();
+      return _numDocs - reduced.getCardinality() - ImmutableRoaringBitmap.andCardinality(reduceInternal(), reduced);
+    } else {
+      if (!other._inverted) {
+        ImmutableRoaringBitmap reduced = reduceInternal();
+        return _numDocs - reduced.getCardinality()
+            + ImmutableRoaringBitmap.andCardinality(other.reduceInternal(), reduced);
+      }
+      return _numDocs - ImmutableRoaringBitmap.andCardinality(reduceInternal(), other.reduceInternal());
+    }
+  }
+
+  private ImmutableRoaringBitmap reduceInternal() {
+    if (_bitmaps.length == 1) {
+      return _bitmaps[0];
+    }
+    return BufferFastAggregation.or(_bitmaps);
+  }
+
+  public ImmutableRoaringBitmap reduce() {
+    if (!_inverted) {
+      return reduceInternal();
+    }
+    return invertedOr();
+  }
+
+  private MutableRoaringBitmap invertedOr() {
```

Comment:
Consider specialize empty and single element `_bitmaps`

Reply:
I don’t think this is necessary.

Reply:
This is unnecessary because when there are no bitmaps, we need to return the entire complement; no bitmaps will be subtracted; the code is correct.